# Model 5.1 and 5.2: Bagging & Random Forest

## Import the libraries and dataset

In [2]:
### Importing the libraries
import numpy as np
import pandas as pd
#from sklearnex import patch_sklearn

#patch_sklearn()

#plotting lib
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns


#Sklearn Lib metrics
from sklearn.metrics import accuracy_score, balanced_accuracy_score
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import GridSearchCV, KFold

# Pipelines : 
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn import set_config


from sklearn.ensemble import RandomForestClassifier

#Missing values : 
from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import KNNImputer, IterativeImputer, SimpleImputer
import missingno as msno

#Dummy
from sklearn.preprocessing import OneHotEncoder, LabelEncoder
from sklearn. preprocessing import StandardScaler


pd.set_option("display.max_columns",None)

In [3]:
train_df = pd.read_csv("Data/train.csv")
test_df = pd.read_csv("Data/test.csv")
train_df.head()

id  occ_code_level2   age  stock_dividends mig_chg_msa  \
0   1                0  42.0              0.0         NaN   
1   2               18  56.0              NaN         NaN   
2   3               26  26.0              NaN         NaN   
3   4                0  67.0              NaN  MSA to MSA   
4   5                0   NaN              NaN    Nonmover   

                 tax_filer_stat                   det_hh_summ  \
0                      Nonfiler                   Householder   
1                           NaN                   Householder   
2           Joint both under 65                   Householder   
3  Joint one under 65 & one 65+                   Householder   
4                      Nonfiler  Child under 18 never married   

  mig_prev_sunbelt hisp_origin             education  wage_per_hour  \
0              NaN   All other            11th grade            0.0   
1              NaN   All other  High school graduate            0.0   
2              NaN   All other  High school graduate            0.0   
3               No   All other                   NaN            0.0   
4  Not in universe   All other              Children            0.0   

   capital_losses     vet_question  own_or_self   country_self mig_move_reg  \
0             NaN  Not in universe            0  United-States          NaN   
1             NaN  Not in universe            2  United-States          NaN   
2             NaN  Not in universe            0          Haiti          NaN   
3             NaN               No            0  United-States  Same county   
4             0.0  Not in universe            0  United-States     Nonmover   

   high_income       hs_college                class_worker  \
0            0  Not in universe             Not in universe   
1            1  Not in universe  Self-employed-incorporated   
2            0  Not in universe                     Private   
3            0  Not in universe             Not in universe   
4            0  Not in universe             Not in universe   

                           mig_same     unemp_reason   state_prev_res  \
0  Not in universe under 1 year old  Not in universe  Not in universe   
1  Not in universe under 1 year old              NaN  Not in universe   
2  Not in universe under 1 year old              NaN  Not in universe   
3                                No              NaN   North Carolina   
4                               Yes  Not in universe  Not in universe   

   ind_code_level2   race country_mother  capital_gains     sex  \
0                0  Black  United-States            0.0  Female   
1               32    NaN  United-States            NaN    Male   
2               41    NaN          Haiti            NaN     NaN   
3                0  Black  United-States            0.0     NaN   
4                0  White  United-States            0.0  Female   

               ind_code_level1                         citizenship  \
0  Not in universe or children   Native- Born in the United States   
1              Wholesale trade   Native- Born in the United States   
2            Hospital services  Foreign born- Not a citizen of U S   
3  Not in universe or children   Native- Born in the United States   
4  Not in universe or children   Native- Born in the United States   

      union_member         fam_under_18                     marital_stat  \
0  Not in universe      Not in universe                              NaN   
1  Not in universe      Not in universe  Married-civilian spouse present   
2  Not in universe      Not in universe                              NaN   
3  Not in universe      Not in universe                              NaN   
4  Not in universe  Mother only present                              NaN   

   region_prev_res  mig_chg_reg country_father  \
0  Not in universe          NaN  United-States   
1  Not in universe          NaN  United-States   
2  Not in universe          NaN          Haiti   
3        Northeast  Same county  United-Stat

## Cleaning the dataset

In [4]:
#removing id for train_df
train_df.drop("id", axis = 1, inplace = True)

In [5]:
# Transforming object into categories 
# for train
for i in train_df.columns:
    if train_df[i].dtypes == "object":
        train_df[i] = train_df[i].astype("category")
# For test
for i in test_df.columns:
    if test_df[i].dtypes == "object":
        test_df[i] = test_df[i].astype("category")

In [6]:
X= train_df.drop("high_income", axis = 1)
y = train_df["high_income"]

In [7]:
# Grouping our features that are categories in one vector
# Same for numeric

categorical_features = [i for i in X.columns if X[i].dtype.name == "category"]
numerical_features = [i for i in X.columns if X[i].dtype.name != "category"]

## Bagging: First try with 1000 trees

In [9]:

numeric_transformer = Pipeline(
    steps=[("imputer_num", SimpleImputer(strategy="mean")),
           ("scaler", StandardScaler())]
)

categorical_transformer = Pipeline(
    steps=[("imputer_cat", SimpleImputer(strategy="constant", fill_value="Missing")),
           ("encoder" , OneHotEncoder(handle_unknown="ignore", sparse=False))]
)



preprocessor = ColumnTransformer(
    transformers=[
        ("num", numeric_transformer, numerical_features),
        ("cat", categorical_transformer, categorical_features),
    ]
)

# Append classifier to preprocessing pipeline.
# Now we have a full prediction pipeline.
clf = Pipeline(
    steps=[("preprocessor", preprocessor), ("classifier", RandomForestClassifier(random_state=1,n_jobs=-1,max_features=None,n_estimators = 1000))]
)


set_config(display="diagram")
clf

Pipeline(steps=[('preprocessor',
                 ColumnTransformer(transformers=[('num',
                                                  Pipeline(steps=[('imputer_num',
                                                                   SimpleImputer()),
                                                                  ('scaler',
                                                                   StandardScaler())]),
                                                  ['occ_code_level2', 'age',
                                                   'stock_dividends',
                                                   'wage_per_hour',
                                                   'capital_losses',
                                                   'own_or_self',
                                                   'ind_code_level2',
                                                   'capital_gains',
                                                   'weeks_worked', 'num_emp',
                                                   'vet_benefits']),
                                                 ('cat',
                                                  Pipeline(steps=[('imputer_ca...
                                                   'class_worker', 'mig_same',
                                                   'unemp_reason',
                                                   'state_prev_res', 'race',
                                                   'country_mother', 'sex',
                                                   'ind_code_level1',
                                                   'citizenship',
                                                   'union_member',
                                                   'fam_under_18',
                                                   'marital_stat',
                                                   'region_prev_res',
                                                   'mig_chg_reg',
                                                   'country_father',
                                                   'occ_code_level1',
                                                   'full_or_part_emp',
                                                   'det_hh_fam_stat'])])),
                ('classifier',
                 RandomForestClassifier(max_features=None, n_estimators=1000,
                                        n_jobs=-1, random_state=1))])

In [10]:
clf.fit(X,y)

Pipeline(steps=[('preprocessor',
                 ColumnTransformer(transformers=[('num',
                                                  Pipeline(steps=[('imputer_num',
                                                                   SimpleImputer()),
                                                                  ('scaler',
                                                                   StandardScaler())]),
                                                  ['occ_code_level2', 'age',
                                                   'stock_dividends',
                                                   'wage_per_hour',
                                                   'capital_losses',
                                                   'own_or_self',
                                                   'ind_code_level2',
                                                   'capital_gains',
                                                   'weeks_worked', 'num_emp',
                                                   'vet_benefits']),
                                                 ('cat',
                                                  Pipeline(steps=[('imputer_ca...
                                                   'class_worker', 'mig_same',
                                                   'unemp_reason',
                                                   'state_prev_res', 'race',
                                                   'country_mother', 'sex',
                                                   'ind_code_level1',
                                                   'citizenship',
                                                   'union_member',
                                                   'fam_under_18',
                                                   'marital_stat',
                                                   'region_prev_res',
                                                   'mig_chg_reg',
                                                   'country_father',
                                                   'occ_code_level1',
                                                   'full_or_part_emp',
                                                   'det_hh_fam_stat'])])),
                ('classifier',
                 RandomForestClassifier(max_features=None, n_estimators=1000,
                                        n_jobs=-1, random_state=1))])

In [13]:
test_id = test_df[["id"]]
test_df.drop("id",axis = 1, inplace = True)

In [15]:
y_test_pred = clf.predict(test_df)
y_test_pred[0:10]

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 1])

In [16]:
test_id["high_income"] = y_test_pred

/var/folders/r3/t8xp3_0d2qg9znqckxzmdc2w0000gp/T/ipykernel_11870/3628057280.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  test_id["high_income"] = y_test_pred


In [17]:
test_id

id  high_income
0        1            0
1        2            0
2        3            0
3        4            0
4        5            0
...    ...          ...
6063  6064            0
6064  6065            1
6065  6066            1
6066  6067            0
6067  6068            1

[6068 rows x 2 columns]

In [18]:
test_id.to_csv("Predictions/bagging_1000_prediction.csv",index = False, header=True)

## Random Forest: First inference on tuning parameters with automatic GridSearch

In [9]:
### Tuning our tree : 
numeric_transformer = Pipeline(
    steps=[("imputer_num", SimpleImputer(strategy="mean")),
           ("scaler", StandardScaler())]
)

categorical_transformer = Pipeline(
    steps=[("imputer_cat", SimpleImputer(strategy="constant", fill_value="Missing")),
           ("encoder" , OneHotEncoder(handle_unknown="ignore", sparse=False))]
)



preprocessor = ColumnTransformer(
    transformers=[
        ("num", numeric_transformer, numerical_features),
        ("cat", categorical_transformer, categorical_features),
    ]
)

# Append classifier to preprocessing pipeline.
# Now we have a full prediction pipeline.
clf = Pipeline(
    steps=[("preprocessor", preprocessor), ("classifier", RandomForestClassifier(random_state=1,n_jobs=-1,max_features="sqrt",n_estimators = 200))]
)


set_config(display="diagram")
clf

Pipeline(steps=[('preprocessor',
                 ColumnTransformer(transformers=[('num',
                                                  Pipeline(steps=[('imputer_num',
                                                                   SimpleImputer()),
                                                                  ('scaler',
                                                                   StandardScaler())]),
                                                  ['occ_code_level2', 'age',
                                                   'stock_dividends',
                                                   'wage_per_hour',
                                                   'capital_losses',
                                                   'own_or_self',
                                                   'ind_code_level2',
                                                   'capital_gains',
                                                   'weeks_worked', 'num_emp',
                                                   'vet_benefits']),
                                                 ('cat',
                                                  Pipeline(steps=[('imputer_ca...
                                                   'class_worker', 'mig_same',
                                                   'unemp_reason',
                                                   'state_prev_res', 'race',
                                                   'country_mother', 'sex',
                                                   'ind_code_level1',
                                                   'citizenship',
                                                   'union_member',
                                                   'fam_under_18',
                                                   'marital_stat',
                                                   'region_prev_res',
                                                   'mig_chg_reg',
                                                   'country_father',
                                                   'occ_code_level1',
                                                   'full_or_part_emp',
                                                   'det_hh_fam_stat'])])),
                ('classifier',
                 RandomForestClassifier(n_estimators=200, n_jobs=-1,
                                        random_state=1))])

Below, we just show the number of dimensions of our predictor space, after performing the feature engineering.

In [12]:
preprocessed_data = pd.DataFrame(preprocessor.fit_transform(X))
preprocessed_data
#410 variables ! 

0             1             2             3         4         5    \
0     -0.827102  3.336430e-01 -1.722629e-01 -2.608265e-01  0.000000 -0.344443   
1      0.472406  1.223516e+00 -2.391037e-17 -2.608265e-01  0.000000  3.097908   
2      1.049965 -6.833552e-01 -2.391037e-17 -2.608265e-01  0.000000 -0.344443   
3     -0.827102  1.922703e+00 -2.391037e-17 -2.608265e-01  0.000000 -0.344443   
4     -0.827102  4.516379e-16 -2.391037e-17 -2.608265e-01 -0.248851 -0.344443   
...         ...           ...           ...           ...       ...       ...   
54602 -0.682712  4.516379e-16 -2.391037e-17 -2.608265e-01  0.000000  3.097908   
54603 -0.827102 -3.019809e-01 -1.722629e-01 -4.994270e-17 -0.248851 -0.344443   
54604 -0.827102 -9.376048e-01 -2.391037e-17 -2.608265e-01  0.000000 -0.344443   
54605 -0.827102 -2.081728e+00 -1.722629e-01 -2.608265e-01 -0.248851 -0.344443   
54606 -0.827102 -2.145290e+00 -1.722629e-01 -2.608265e-01 -0.248851 -0.344443   

            6         7             8         9         10   11   12   13   \
0     -1.026691 -0.194292 -1.531480e+00 -1.008107  0.476955  0.0  0.0  0.0   
1      0.712314  0.000000 -3.875967e-16 -0.603743  0.476955  0.0  0.0  0.0   
2      1.201410  0.000000 -3.875967e-16  0.204985  0.476955  0.0  0.0  0.0   
3     -1.026691 -0.194292 -1.531480e+00 -1.008107 -0.830922  0.0  0.0  1.0   
4     -1.026691 -0.194292 -3.875967e-16 -1.008107 -2.138799  0.0  0.0  0.0   
...         ...       ...           ...       ...       ...  ...  ...  ...   
54602  0.331907  0.000000 -3.875967e-16  0.609349  0.476955  0.0  0.0  0.0   
54603 -1.026691 -0.194292 -1.531480e+00 -1.008107  0.476955  0.0  0.0  0.0   
54604 -1.026691  0.000000 -8.768872e-01 -0.603743  0.476955  0.0  0.0  0.0   
54605 -1.026691 -0.194292 -1.531480e+00 -1.008107 -2.138799  0.0  0.0  0.0   
54606 -1.026691  0.000000 -1.531480e+00 -1.008107 -2.138799  0.0  0.0  0.0   

       14   15   16   17   18   19   20   21   22   23   24   25   26   27   \
0      0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   
1      0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0   
2      0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0   
3      0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0   
4      0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   
...    ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
54602  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0   
54603  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   
54604  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0   
54605  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   
54606  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   

       28   29   30   31   32   33   34   35   36   37   38   39   40   41   \
0      0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0  0.0   
1      0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0  0.0   
2      0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0  0.0   
3      0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0  0.0   
4      0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  1.0  0.0   
...    ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
54602  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  1.0  0.0   
54603  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0  0.0   
54604  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  1.0  0.0   
54605  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0  0.0   
54606  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0  0.0   

       42   43   44   45   46   47   48   49   50   51   52   53   54   55   \
0      0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0   
1      0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 

In [14]:
int(np.sqrt(len(preprocessed_data.columns)))

20

**COMMENT**

As seen in class, it is common to use $m \approx \sqrt{p}$. THis is why, we will perform a GridSearch around this number of features.
Thus, we can perform a first automatic GridSearch for Random Forest:

In [20]:
hyper_param = {"classifier__max_features" : np.arange(15,25),
              "classifier__min_samples_leaf": np.arange(1,6,2)}
hyper_param

{'classifier__max_features': array([15, 16, 17, 18, 19, 20, 21, 22, 23, 24]),
 'classifier__min_samples_leaf': array([1, 3, 5])}

In [22]:
rf_grid_search = GridSearchCV(estimator = clf,param_grid=hyper_param, scoring="accuracy",
                           cv = KFold(n_splits=5, shuffle=True, random_state=1), n_jobs=-1, verbose=1)

In [23]:
rf_grid_search.fit(X,y)

Fitting 5 folds for each of 30 candidates, totalling 150 fits


GridSearchCV(cv=KFold(n_splits=5, random_state=1, shuffle=True),
             estimator=Pipeline(steps=[('preprocessor',
                                        ColumnTransformer(transformers=[('num',
                                                                         Pipeline(steps=[('imputer_num',
                                                                                          SimpleImputer()),
                                                                                         ('scaler',
                                                                                          StandardScaler())]),
                                                                         ['occ_code_level2',
                                                                          'age',
                                                                          'stock_dividends',
                                                                          'wage_per_hour',
                                                                          'capital_losses',
                                                                          'own_or_self',
                                                                          'ind_code_level2',
                                                                          'capital_gains',
                                                                          'week...
                                                                          'mig_chg_reg',
                                                                          'country_father',
                                                                          'occ_code_level1',
                                                                          'full_or_part_emp',
                                                                          'det_hh_fam_stat'])])),
                                       ('classifier',
                                        RandomForestClassifier(max_features='sqrt',
                                                               n_estimators=200,
                                                               n_jobs=-1,
                                                               random_state=1))]),
             n_jobs=-1,
             param_grid={'classifier__max_features': array([15, 16, 17, 18, 19, 20, 21, 22, 23, 24]),
                         'classifier__min_samples_leaf': array([1, 3, 5])},
             scoring='accuracy', verbose=1)

In [51]:
cv_res_new = pd.DataFrame(rf_grid_search.cv_results_)
cv_res_new

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
0       55.925575      1.338859         2.463253        0.184390   
1       49.666907      4.867348         2.394748        0.419958   
2       38.090263      1.854731         2.082220        0.406755   
3       54.741662      1.804889         2.814968        0.529073   
4       44.907965      0.634381         2.395191        0.317428   
5       41.650819      1.374164         2.217968        0.170026   
6       55.189792      1.165297         2.843550        0.169019   
7       46.112147      0.798703         2.623310        0.244764   
8       42.224755      1.153197         2.499745        0.215450   
9       64.241735      3.332654         3.511349        0.748768   
10      52.741606      3.916471         2.623898        0.131573   
11      43.363033      0.558115         2.285134        0.178726   
12      68.008506      1.778331         3.308798        0.783421   
13      59.630424      5.831941         2.683585        0.407852   
14      62.086443      1.535923         2.547162        0.206052   
15      67.318464      4.829666         2.862593        0.268840   
16      69.766383      3.894241         3.811357        0.807279   
17      59.187363      4.056549         2.559035        0.291352   
18      83.787119      9.696895         4.146263        0.671900   
19      86.903389      3.394540         3.427945        0.342600   
20      64.689351      3.152831         2.887936        0.347925   
21      78.762054      4.709323         3.433559        0.354900   
22      64.179323      1.412165         3.106985        0.436523   
23      66.253233      3.007495         3.017203        0.365591   
24      80.796262      1.631508         3.255230        0.276132   
25      69.171181      1.278823         3.295754        0.684175   
26      64.983887      2.748176         3.098210        0.433941   
27      84.868723      3.427135         3.143579        0.105145   
28      74.592022      5.723047         3.310701        0.422710   
29      64.390767     13.580969         2.151792        0.991016   

   param_classifier__max_features param_classifier__min_samples_leaf  \
0                              15                                  1   
1                              15                                  3   
2                              15                                  5   
3                              16                                  1   
4                              16                                  3   
5                              16                                  5   
6                              17                                  1   
7                              17                                  3   
8                              17                                  5   
9                              18                                  1   
10                             18                                  3   
11                             18                                  5   
12                             19                                  1   
13                             19                                  3   
14                             19                                  5   
15                             20                                  1   
16                             20                                  3   
17                             20                                  5   
18                             21                                  1   
19                             21                                  3   
20                             21                                  5   
21                             22                                  1   
22                             22                                  3   
23                             22                                  5   
24                             23                                  1   
25                     

In [59]:
mean_scores = cv_res.cv_results_["mean_test_score"]
best_index = np.argmax(mean_scores)
se_scores = cv_res.cv_results_["std_test_score"] / np.sqrt(5)


one_stand_error_data_frame = cv_res_new[cv_res_new["mean_test_score"] >= (mean_scores[best_index] - se_scores[best_index])]
one_stand_error_data_frame[one_stand_error_data_frame["mean_test_score"] == one_stand_error_data_frame["mean_test_score"].min()]

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
18      83.787119      9.696895         4.146263          0.6719   

   param_classifier__max_features param_classifier__min_samples_leaf  \
18                             21                                  1   

                                               params  split0_test_score  \
18  {'classifier__max_features': 21, 'classifier__...           0.862113   

    split1_test_score  split2_test_score  split3_test_score  \
18           0.862754            0.85679           0.858255   

    split4_test_score  mean_test_score  std_test_score  rank_test_score  
18           0.858987          0.85978        0.002288                5

**COMMENT**

It seems that the best model according to the 1 standard-error rule has 21 features and min_samples_leaf of 1.

In [60]:
clf.set_params(classifier__max_features= 21, classifier__min_samples_leaf = 1, classifier__n_estimators = 600)

Pipeline(steps=[('preprocessor',
                 ColumnTransformer(transformers=[('num',
                                                  Pipeline(steps=[('imputer_num',
                                                                   SimpleImputer()),
                                                                  ('scaler',
                                                                   StandardScaler())]),
                                                  ['occ_code_level2', 'age',
                                                   'stock_dividends',
                                                   'wage_per_hour',
                                                   'capital_losses',
                                                   'own_or_self',
                                                   'ind_code_level2',
                                                   'capital_gains',
                                                   'weeks_worked', 'num_emp',
                                                   'vet_benefits']),
                                                 ('cat',
                                                  Pipeline(steps=[('imputer_ca...
                                                   'class_worker', 'mig_same',
                                                   'unemp_reason',
                                                   'state_prev_res', 'race',
                                                   'country_mother', 'sex',
                                                   'ind_code_level1',
                                                   'citizenship',
                                                   'union_member',
                                                   'fam_under_18',
                                                   'marital_stat',
                                                   'region_prev_res',
                                                   'mig_chg_reg',
                                                   'country_father',
                                                   'occ_code_level1',
                                                   'full_or_part_emp',
                                                   'det_hh_fam_stat'])])),
                ('classifier',
                 RandomForestClassifier(max_features=21, n_estimators=600,
                                        n_jobs=-1, random_state=1))])

In [61]:
clf.fit(X,y)

Pipeline(steps=[('preprocessor',
                 ColumnTransformer(transformers=[('num',
                                                  Pipeline(steps=[('imputer_num',
                                                                   SimpleImputer()),
                                                                  ('scaler',
                                                                   StandardScaler())]),
                                                  ['occ_code_level2', 'age',
                                                   'stock_dividends',
                                                   'wage_per_hour',
                                                   'capital_losses',
                                                   'own_or_self',
                                                   'ind_code_level2',
                                                   'capital_gains',
                                                   'weeks_worked', 'num_emp',
                                                   'vet_benefits']),
                                                 ('cat',
                                                  Pipeline(steps=[('imputer_ca...
                                                   'class_worker', 'mig_same',
                                                   'unemp_reason',
                                                   'state_prev_res', 'race',
                                                   'country_mother', 'sex',
                                                   'ind_code_level1',
                                                   'citizenship',
                                                   'union_member',
                                                   'fam_under_18',
                                                   'marital_stat',
                                                   'region_prev_res',
                                                   'mig_chg_reg',
                                                   'country_father',
                                                   'occ_code_level1',
                                                   'full_or_part_emp',
                                                   'det_hh_fam_stat'])])),
                ('classifier',
                 RandomForestClassifier(max_features=21, n_estimators=600,
                                        n_jobs=-1, random_state=1))])

### Export the predictions to .csv

In [62]:
y_train_pred = clf.predict(X)
y_train_pred

array([0, 1, 0, ..., 0, 0, 0])

In [65]:
new_tune_200_tree = test_df[["id"]]
new_tune_200_tree

id
0        1
1        2
2        3
3        4
4        5
...    ...
6063  6064
6064  6065
6065  6066
6066  6067
6067  6068

[6068 rows x 1 columns]

In [66]:
y_test_pred = clf.predict(test_df.drop("id",axis = 1))
y_test_pred

array([0, 0, 0, ..., 0, 0, 1])

In [67]:
new_tune_200_tree["high_income"] = y_test_pred
new_tune_200_tree

/var/folders/r3/t8xp3_0d2qg9znqckxzmdc2w0000gp/T/ipykernel_16032/2699741353.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_tune_200_tree["high_income"] = y_test_pred


id  high_income
0        1            0
1        2            0
2        3            0
3        4            0
4        5            0
...    ...          ...
6063  6064            0
6064  6065            1
6065  6066            0
6066  6067            0
6067  6068            1

[6068 rows x 2 columns]

In [68]:
new_tune_200_tree.to_csv("Predictions/rf_24features_200_prediction.csv",index = False, header=True)

## Random Forest: Trial with 2000 trees

The number of trees is not a tuning parameter. We can increase it's number to decrease the bias !

In [14]:
hyper_param = {"classifier__max_features" : [21],
              "classifier__min_samples_leaf": [1],
              "classifier__n_estimators":[2000]}
rf_grid_search = GridSearchCV(estimator = clf,param_grid=hyper_param, scoring="accuracy",
                           cv = KFold(n_splits=5, shuffle=True, random_state=1), n_jobs=-1, verbose=1)

In [15]:
rf_grid_search.fit(X,y)

Fitting 5 folds for each of 1 candidates, totalling 5 fits


GridSearchCV(cv=KFold(n_splits=5, random_state=1, shuffle=True),
             estimator=Pipeline(steps=[('preprocessor',
                                        ColumnTransformer(transformers=[('num',
                                                                         Pipeline(steps=[('imputer_num',
                                                                                          SimpleImputer()),
                                                                                         ('scaler',
                                                                                          StandardScaler())]),
                                                                         ['occ_code_level2',
                                                                          'age',
                                                                          'stock_dividends',
                                                                          'wage_per_hour',
                                                                          'capital_losses',
                                                                          'own_or_self',
                                                                          'ind_code_level2',
                                                                          'capital_gains',
                                                                          'week...
                                                                          'marital_stat',
                                                                          'region_prev_res',
                                                                          'mig_chg_reg',
                                                                          'country_father',
                                                                          'occ_code_level1',
                                                                          'full_or_part_emp',
                                                                          'det_hh_fam_stat'])])),
                                       ('classifier',
                                        RandomForestClassifier(n_estimators=200,
                                                               n_jobs=-1,
                                                               random_state=1))]),
             n_jobs=-1,
             param_grid={'classifier__max_features': [21],
                         'classifier__min_samples_leaf': [1],
                         'classifier__n_estimators': [2000]},
             scoring='accuracy', verbose=1)

In [16]:
pd.DataFrame(rf_grid_search.cv_results_)

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
0     580.496705    216.264528        16.991444        6.597619   

  param_classifier__max_features param_classifier__min_samples_leaf  \
0                             21                                  1   

  param_classifier__n_estimators  \
0                           2000   

                                              params  split0_test_score  \
0  {'classifier__max_features': 21, 'classifier__...           0.862754   

   split1_test_score  split2_test_score  split3_test_score  split4_test_score  \
0           0.862846           0.855508           0.860086           0.858896   

   mean_test_score  std_test_score  rank_test_score  
0         0.860018        0.002724                1

We are havin a mean score of 0.860018 which is the best so far ! 

In [9]:
clf.set_params(classifier__max_features= 21, classifier__min_samples_leaf = 1, classifier__n_estimators = 2000)

Pipeline(steps=[('preprocessor',
                 ColumnTransformer(transformers=[('num',
                                                  Pipeline(steps=[('imputer_num',
                                                                   SimpleImputer()),
                                                                  ('scaler',
                                                                   StandardScaler())]),
                                                  ['occ_code_level2', 'age',
                                                   'stock_dividends',
                                                   'wage_per_hour',
                                                   'capital_losses',
                                                   'own_or_self',
                                                   'ind_code_level2',
                                                   'capital_gains',
                                                   'weeks_worked', 'num_emp',
                                                   'vet_benefits']),
                                                 ('cat',
                                                  Pipeline(steps=[('imputer_ca...
                                                   'class_worker', 'mig_same',
                                                   'unemp_reason',
                                                   'state_prev_res', 'race',
                                                   'country_mother', 'sex',
                                                   'ind_code_level1',
                                                   'citizenship',
                                                   'union_member',
                                                   'fam_under_18',
                                                   'marital_stat',
                                                   'region_prev_res',
                                                   'mig_chg_reg',
                                                   'country_father',
                                                   'occ_code_level1',
                                                   'full_or_part_emp',
                                                   'det_hh_fam_stat'])])),
                ('classifier',
                 RandomForestClassifier(max_features=21, n_estimators=2000,
                                        n_jobs=-1, random_state=1))])

In [10]:
clf.fit(X,y)

Pipeline(steps=[('preprocessor',
                 ColumnTransformer(transformers=[('num',
                                                  Pipeline(steps=[('imputer_num',
                                                                   SimpleImputer()),
                                                                  ('scaler',
                                                                   StandardScaler())]),
                                                  ['occ_code_level2', 'age',
                                                   'stock_dividends',
                                                   'wage_per_hour',
                                                   'capital_losses',
                                                   'own_or_self',
                                                   'ind_code_level2',
                                                   'capital_gains',
                                                   'weeks_worked', 'num_emp',
                                                   'vet_benefits']),
                                                 ('cat',
                                                  Pipeline(steps=[('imputer_ca...
                                                   'class_worker', 'mig_same',
                                                   'unemp_reason',
                                                   'state_prev_res', 'race',
                                                   'country_mother', 'sex',
                                                   'ind_code_level1',
                                                   'citizenship',
                                                   'union_member',
                                                   'fam_under_18',
                                                   'marital_stat',
                                                   'region_prev_res',
                                                   'mig_chg_reg',
                                                   'country_father',
                                                   'occ_code_level1',
                                                   'full_or_part_emp',
                                                   'det_hh_fam_stat'])])),
                ('classifier',
                 RandomForestClassifier(max_features=21, n_estimators=2000,
                                        n_jobs=-1, random_state=1))])

### Export the predictions to .csv

In [11]:
new_tune_2000_tree = test_df[["id"]]
new_tune_2000_tree

id
0        1
1        2
2        3
3        4
4        5
...    ...
6063  6064
6064  6065
6065  6066
6066  6067
6067  6068

[6068 rows x 1 columns]

In [12]:
y_test_pred = clf.predict(test_df.drop("id",axis = 1))
y_test_pred

array([0, 0, 0, ..., 0, 0, 1])

In [13]:
new_tune_2000_tree["high_income"] = y_test_pred
new_tune_2000_tree

/var/folders/r3/t8xp3_0d2qg9znqckxzmdc2w0000gp/T/ipykernel_23702/1942914794.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_tune_2000_tree["high_income"] = y_test_pred


id  high_income
0        1            0
1        2            0
2        3            0
3        4            0
4        5            0
...    ...          ...
6063  6064            0
6064  6065            1
6065  6066            0
6066  6067            0
6067  6068            1

[6068 rows x 2 columns]

In [14]:
new_tune_2000_tree.to_csv("Predictions/rf_21features_2000_prediction.csv",index = False, header=True)